### DISCUSSION TITLE
Pre-condition Violation when using HasReactionSubstructMatch

### DISCUSSION THREAD

**Discussion by victorsabanza** [rdkit/rdkit#6345] [General]
Hi, 

I'm trying to use `HasReactionSubstructMatch` from the `Chem.rdChemReactions` module to check if a reaction SMILES matches a given reaction SMARTS. I'm using this code:
```
smarts = '[c:1][O;D2:2][CH2:3]>>[c:1][OH:2].[CH3:3]'

rxn = AllChem.ReactionFromSmarts(smarts)

rxn_smiles = 'c1ccccc1OCCCC>>c1ccccc1O.CCCC'

rxn_smiles = AllChem.ReactionFromSmarts(rxn_smiles, useSmiles=True)

HasReactionSubstructMatch(rxn_smiles, rxn)
```
However, I get this error:
```
RuntimeError: Pre-condition Violation
	getNumImplicitHs() called without preceding call to calcImplicitValence()
	Violation occurred on line 297 in file Code/GraphMol/Atom.cpp
	Failed Expression: d_implicitValence > -1
	RDKIT: 2022.09.5
	BOOST: 1_78
```
Does anyone know what could be wrong? Thanks in advance!

**Comment by mainguyenanhvu** [comment]
Hi, do you fix it?

**Comment by greglandrum** [comment]
The problem here is that no sanitization is done when a reaction is parsed; this means that no valences or implicit H counts are calculated. The SMARTS for your query reaction includes implicit H queries, so it cannot work without sanitization.

The solution is to sanitize the reaction you create from SMILES. Then the substructure query works as expected:
```
In [13]: rdChemReactions.SanitizeRxn(rxn_smiles)
Out[13]: rdkit.Chem.rdChemReactions.SanitizeFlags.SANITIZE_NONE

In [14]: rdChemReactions.HasReactionSubstructMatch(rxn_smiles, rxn)
Out[14]: True
```
